compiler/rustc_mir_dataflow/src/impls/liveness.rs

```diff
@@ -74,22 +131,30 @@ impl<'tcx> GenKillAnalysis<'tcx> for MaybeLiveLocals {
         return_places: CallReturnPlaces<'_, 'tcx>,
     ) {
         if let CallReturnPlaces::Yield(resume_place) = return_places {
-            YieldResumeEffect(trans).visit_place(
+            YieldResumeEffect { trans, upvars: self.upvars }.visit_place(
                 &resume_place,
                 PlaceContext::MutatingUse(MutatingUseContext::Yield),
                 Location::START,
             )
         } else {
+            trace!(?return_places, "(return) liveness kill?");
             return_places.for_each(|place| {
                 if let Some(local) = place.as_local() {
                     trans.kill(local);
+                } else if place.local == ty::CAPTURE_STRUCT_LOCAL
+                    && let [PlaceElem::Field(..)] = &**place.projection
+                {
+                    maybe_upvar_kill(trans, &place, self.upvars, Location::START);
                 }
             });
         }
     }
 }
 
-pub struct TransferFunction<'a, T>(pub &'a mut T);
+pub struct TransferFunction<'a, T> {
+    pub trans: &'a mut T,
+    pub upvars: Option<(Local, usize)>,
```

Comment:
I don't remember if earlier iterations on this PR had these changes to the various dataflow impls; its possible they did and I just overlooked them.

But seeing them now with fresh eyes, it just seems unfortunate that so many dataflow implementations seem like they are expected to duplicate this `upvars: Option<(Local, usize)>` state pattern. I would be very concerned about whether it is a footgun to forget to do it, and also whether it makes all of the dataflow systems harder to maintain.

Having said that, I also don't have a better suggestion about what to do for the short term. It probably is something to bring up with the broader team, in terms of explaining why the pattern is necessary and seeing if anyone has ideas for another way to accomplish the objective.